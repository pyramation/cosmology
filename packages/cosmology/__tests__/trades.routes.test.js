// @ts-nocheck
import validatorPricesFixture from '../__fixtures__/validator/tokens/v2/all/data.json';
import pairsFixture from '../__fixtures__/validator/pairs/v1/summary/data.json';
import poolsFixture from '../__fixtures__/lcd/osmosis/gamm/v1beta1/pools/data.json';
import lockedPoolsFixture from '../__fixtures__/lcd/osmosis/lockup/v1beta1/account_locked_coins/osmo1/data.json';

import {
  symbolsAndDisplayValuesToCoinsArray,
  getTradesRequiredToGetBalances,
  getSwaps,
  getFilteredPoolsWithValues,
  convertValidatorPricesToDenomPriceHash
} from '../src/utils/osmo';

const prices = convertValidatorPricesToDenomPriceHash(validatorPricesFixture);
const pools = getFilteredPoolsWithValues({ prices, pools: poolsFixture.pools });

it('single route swaps', async () => {
  const balances = symbolsAndDisplayValuesToCoinsArray([
    {
      symbol: 'ATOM',
      amount: 1000
    },
    {
      symbol: 'OSMO',
      amount: 1000
    }
  ]);
  const desired = symbolsAndDisplayValuesToCoinsArray([
    {
      symbol: 'STARS',
      amount: 1000
    },
    {
      symbol: 'CMDX',
      amount: 100
    },
    {
      symbol: 'HUAHUA',
      amount: 10000
    }
  ]);
  const trades = getTradesRequiredToGetBalances({ prices, balances, desired });

  console.log(trades);

  const swaps = getSwaps({
    prices,
    pools,
    trades,
    pairs: pairsFixture.data
  });
  swaps.forEach((swap) => {
    expect(swap.routes.length).toBe(1);
  });
  expect(swaps).toMatchSnapshot();
});

// {
//     "type": "osmosis/gamm/swap-exact-amount-in",
//     "value": {
//       "sender": "osmo1mn7ltv80e0vnr9zsy2ukuxaqk0yr4n9em87wsf",
//       "routes": [
//         {
//           "poolId": "611",
//           "tokenOutDenom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2"
//         },
//         {
//           "poolId": "600",
//           "tokenOutDenom": "ibc/EA3E1640F9B1532AB129A571203A0B9F789A7F14BB66E350DCBFA18E1A1931F0"
//         }
//       ],
//       "tokenIn": {
//         "denom": "ibc/987C17B11ABC2B20019178ACE62929FE9840202CE79498E29FE8E5CB02B7C0A4",
//         "amount": "1000000"
//       },
//       "tokenOutMinAmount": "268878"
//     }
//   }

it('multi-hop route swaps', async () => {
  const balances = symbolsAndDisplayValuesToCoinsArray([
    {
      symbol: 'STARS',
      amount: 1000000
    },
    {
      symbol: 'HUAHUA',
      amount: 10000000000
    }
  ]);
  const desired = symbolsAndDisplayValuesToCoinsArray([
    {
      symbol: 'CMDX',
      amount: 100
    },
    {
      symbol: 'LUNA',
      amount: 100
    }
  ]);
  const trades = getTradesRequiredToGetBalances({ prices, balances, desired });

  console.log(trades);

  const swaps = getSwaps({
    prices,
    pools,
    trades,
    pairs: pairsFixture.data
  });
  swaps.forEach((swap) => {
    expect(swap.routes.length).toBe(2);
  });
  expect(swaps).toMatchSnapshot();
});
