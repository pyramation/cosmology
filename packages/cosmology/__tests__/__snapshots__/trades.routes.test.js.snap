// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multi-hop route swaps 1`] = `
Array [
  Object {
    "routes": Array [
      Object {
        "liquidity": 31502013.872773834,
        "poolId": "611",
        "tokenInSymbol": "STARS",
        "tokenOutDenom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
        "tokenOutSymbol": "ATOM",
      },
      Object {
        "liquidity": 14724586.80624463,
        "poolId": "600",
        "tokenInSymbol": "ATOM",
        "tokenOutDenom": "ibc/EA3E1640F9B1532AB129A571203A0B9F789A7F14BB66E350DCBFA18E1A1931F0",
        "tokenOutSymbol": "CMDX",
      },
    ],
    "trade": Object {
      "beliefValue": "265.049267790000000000",
      "buy": Object {
        "amount": "100000000.0000000000000000",
        "denom": "ibc/EA3E1640F9B1532AB129A571203A0B9F789A7F14BB66E350DCBFA18E1A1931F0",
        "displayAmount": "100.000000000000000000",
        "symbol": "CMDX",
        "value": "265.049267790000000000",
      },
      "sell": Object {
        "amount": "371027791.0289713471570000",
        "denom": "ibc/987C17B11ABC2B20019178ACE62929FE9840202CE79498E29FE8E5CB02B7C0A4",
        "displayAmount": "371.027791028971347157",
        "symbol": "STARS",
        "value": "265.049267790000000000",
      },
    },
  },
  Object {
    "routes": Array [
      Object {
        "liquidity": 31502013.872773834,
        "poolId": "611",
        "tokenInSymbol": "STARS",
        "tokenOutDenom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
        "tokenOutSymbol": "ATOM",
      },
      Object {
        "liquidity": 5022123.40019734,
        "poolId": "565",
        "tokenInSymbol": "ATOM",
        "tokenOutDenom": "ibc/0EF15DF2F02480ADE0BB6E85D9EBB5DAEA2836D3860E9F97F9AADE4F57A31AA0",
        "tokenOutSymbol": "LUNA",
      },
    ],
    "trade": Object {
      "beliefValue": "5270.000000000000000000",
      "buy": Object {
        "amount": "100000000.0000000000000000",
        "denom": "ibc/0EF15DF2F02480ADE0BB6E85D9EBB5DAEA2836D3860E9F97F9AADE4F57A31AA0",
        "displayAmount": "100.000000000000000000",
        "symbol": "LUNA",
        "value": "5270.000000000000000000",
      },
      "sell": Object {
        "amount": "7377181137.0250116604400000",
        "denom": "ibc/987C17B11ABC2B20019178ACE62929FE9840202CE79498E29FE8E5CB02B7C0A4",
        "displayAmount": "7377.181137025011660440",
        "symbol": "STARS",
        "value": "5270.000000000000000000",
      },
    },
  },
]
`;

exports[`single route swaps 1`] = `
Array [
  Object {
    "routes": Array [
      Object {
        "liquidity": 31502013.872773834,
        "poolId": "611",
        "tokenInSymbol": "ATOM",
        "tokenOutDenom": "ibc/987C17B11ABC2B20019178ACE62929FE9840202CE79498E29FE8E5CB02B7C0A4",
        "tokenOutSymbol": "STARS",
      },
    ],
    "trade": Object {
      "beliefValue": "714.364999600000000000",
      "buy": Object {
        "amount": "1000000000.0000000000000000",
        "denom": "ibc/987C17B11ABC2B20019178ACE62929FE9840202CE79498E29FE8E5CB02B7C0A4",
        "displayAmount": "1000.000000000000000000",
        "symbol": "STARS",
        "value": "714.364999600000000000",
      },
      "sell": Object {
        "amount": "24908124.1143654114370000",
        "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
        "displayAmount": "24.908124114365411437",
        "symbol": "ATOM",
        "value": "714.364999600000000000",
      },
    },
  },
  Object {
    "routes": Array [
      Object {
        "liquidity": 14724586.80624463,
        "poolId": "600",
        "tokenInSymbol": "ATOM",
        "tokenOutDenom": "ibc/EA3E1640F9B1532AB129A571203A0B9F789A7F14BB66E350DCBFA18E1A1931F0",
        "tokenOutSymbol": "CMDX",
      },
    ],
    "trade": Object {
      "beliefValue": "265.049267790000000000",
      "buy": Object {
        "amount": "100000000.0000000000000000",
        "denom": "ibc/EA3E1640F9B1532AB129A571203A0B9F789A7F14BB66E350DCBFA18E1A1931F0",
        "displayAmount": "100.000000000000000000",
        "symbol": "CMDX",
        "value": "265.049267790000000000",
      },
      "sell": Object {
        "amount": "9241606.2688284518830000",
        "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
        "displayAmount": "9.241606268828451883",
        "symbol": "ATOM",
        "value": "265.049267790000000000",
      },
    },
  },
  Object {
    "routes": Array [
      Object {
        "liquidity": 10147526.360461282,
        "poolId": "606",
        "tokenInSymbol": "ATOM",
        "tokenOutDenom": "ibc/B9E0A1A524E98BB407D3CED8720EFEFD186002F90C1B1B7964811DD0CCC12228",
        "tokenOutSymbol": "HUAHUA",
      },
    ],
    "trade": Object {
      "beliefValue": "43.336702000000000000",
      "buy": Object {
        "amount": "10000000000.0000000000000000",
        "denom": "ibc/B9E0A1A524E98BB407D3CED8720EFEFD186002F90C1B1B7964811DD0CCC12228",
        "displayAmount": "10000.000000000000000000",
        "symbol": "HUAHUA",
        "value": "43.336702000000000000",
      },
      "sell": Object {
        "amount": "1511042.6080892608090000",
        "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
        "displayAmount": "1.511042608089260809",
        "symbol": "ATOM",
        "value": "43.336702000000000000",
      },
    },
  },
]
`;
